### PR TITLE
Make use of Parquet size statistics when reading

### PR DESCRIPTION
**Pull request by etseidl**
## Description
#14000 added writing of page size statistics to the Parquet writer.  This PR is an attempt to use those statistics to skip some preprocessing steps, primarily the page string size calculations.

Still a work-in-progress...feedback welcome.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by etseidl** [comment]
Started some perf testing, and found that reading the indexes was having a negative impact on the non-string benchmarks. After dealing with that, the current code is anywhere from 1-20% faster for strings.
```
# parquet_read_decode

## [0] NVIDIA RTX A6000

|  data_type  |    io_type    |  cardinality  |  run_length  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |         Diff |   %Diff |  
|-------------|---------------|---------------|--------------|------------|-------------|------------|-------------|--------------|---------|
|  INTEGRAL   | DEVICE_BUFFER |       0       |      1       |  13.827 ms |       0.74% |  13.950 ms |       0.81% |   122.719 us |   0.89% |   
|  INTEGRAL   | DEVICE_BUFFER |     1000      |      1       |  12.945 ms |       0.39% |  12.989 ms |       0.46% |    44.126 us |   0.34% |   
|  INTEGRAL   | DEVICE_BUFFER |       0       |      32      |  11.600 ms |       0.34% |  11.613 ms |       0.37% |    13.641 us |   0.12% |   
|  INTEGRAL   | DEVICE_BUFFER |     1000      |      32      |  11.043 ms |       0.70% |  11.014 ms |       0.50% |   -28.921 us |  -0.26% |   
|   STRING    | DEVICE_BUFFER |       0       |      1       |  46.105 ms |       0.48% |  45.537 ms |       0.45% |  -567.535 us |  -1.23% |   
|   STRING    | DEVICE_BUFFER |     1000      |      1       |  13.635 ms |       0.72% |  12.844 ms |       0.77% |  -790.722 us |  -5.80% |   
|   STRING    | DEVICE_BUFFER |       0       |      32      |  46.278 ms |       0.49% |  45.663 ms |       0.44% |  -615.361 us |  -1.33% |   
|   STRING    | DEVICE_BUFFER |     1000      |      32      |  12.796 ms |       1.30% |  10.352 ms |       1.49% | -2444.198 us | -19.10% |   
|    LIST     | DEVICE_BUFFER |       0       |      1       |  75.977 ms |       0.50% |  75.457 ms |       0.14% |  -520.594 us |  -0.69% |   
|    LIST     | DEVICE_BUFFER |     1000      |      1       |  63.353 ms |       0.45% |  62.867 ms |       0.48% |  -485.231 us |  -0.77% |   
|    LIST     | DEVICE_BUFFER |       0       |      32      |  53.426 ms |       1.04% |  53.189 ms |       1.19% |  -236.433 us |  -0.44% |   
|    LIST     | DEVICE_BUFFER |     1000      |      32      |  54.762 ms |       1.00% |  54.658 ms |       1.03% |  -103.946 us |  -0.19% |   
|   STRUCT    | DEVICE_BUFFER |       0       |      1       |  53.255 ms |       2.06% |  53.375 ms |       2.04% |   119.537 us |   0.22% |   
|   STRUCT    | DEVICE_BUFFER |     1000      |      1       |  32.553 ms |       0.43% |  32.295 ms |       0.67% |  -257.358 us |  -0.79% |   
|   STRUCT    | DEVICE_BUFFER |       0       |      32      |  52.191 ms |       2.68% |  52.075 ms |       2.62% |  -116.575 us |  -0.22% |   
|   STRUCT    | DEVICE_BUFFER |     1000      |      32      |  30.491 ms |       0.22% |  29.165 ms |       0.41% | -1325.684 us |  -4.35% | 
```

**Comment by etseidl** [comment]
For comparison, reading the page stats when they aren't used has the following impact:
```
|  data_type  |    io_type    |  cardinality  |  run_length  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |         Diff |   %Diff |
|-------------|---------------|---------------|--------------|------------|-------------|------------|-------------|--------------|---------|
|  INTEGRAL   | DEVICE_BUFFER |       0       |      1       |  13.827 ms |       0.74% |  15.025 ms |       2.35% |     1.198 ms |   8.66% |
|  INTEGRAL   | DEVICE_BUFFER |     1000      |      1       |  12.945 ms |       0.39% |  13.954 ms |       1.06% |     1.009 ms |   7.79% |
|  INTEGRAL   | DEVICE_BUFFER |       0       |      32      |  11.600 ms |       0.34% |  12.603 ms |       0.64% |     1.003 ms |   8.65% |
|  INTEGRAL   | DEVICE_BUFFER |     1000      |      32      |  11.043 ms |       0.70% |  12.062 ms |       0.36% |     1.019 ms |   9.23% |
```

File: cpp/src/io/parquet/reader_impl_chunking.cu [hunk 1]
```diff
@@ -359,6 +359,28 @@ void reader::impl::create_global_chunk_info()
...
+    auto const& rg      = row_groups_info[0];
+    auto const& columns = _metadata->get_row_group(rg.index, rg.source_index).columns;
+    column_mapping.resize(num_input_columns);
+    std::transform(
+      _input_columns.begin(), _input_columns.end(), column_mapping.begin(), [&](auto const& col) {
+        // translate schema_idx into something we can use for the page indexes
+        if (auto it = std::find_if(
+              columns.begin(),
+              columns.end(),
+              [&col](auto const& col_chunk) { return col_chunk.schema_idx == col.schema_idx; });
```
Comment (vuule) [inline cpp/src/io/parquet/reader_impl_chunking.cu:376]: are columns maybe sorted by schema_idx?

**Review by vuule — COMMENTED** [review]
No fundamental concerns with the changes, just some minor comments/questions.

File: cpp/src/io/parquet/reader_impl_chunking.cu (hunk 1 above)
Comment (etseidl) [reply cpp/src/io/parquet/reader_impl_chunking.cu:376]: The input columns can be in any order depending on user input. The `schema_idx` is determined by the file schema IIRC.

File: cpp/src/io/parquet/reader_impl_helpers.hpp [hunk 2]
```diff
@@ -43,12 +80,20 @@ struct row_group_info {
   size_t start_row;
   size_type source_index;  // file index.
 
+  // Optional metadata pulled from the column and offset indexes, if present.
+  std::optional<std::vector<column_info>> columns;
```
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/reader_impl_helpers.hpp:84]: Naming is my kryptonite. Could be `column_chunk_info` to be more in line with `row_group_info`. Likewise `columns` can be `column_chunks`.

File: cpp/src/io/parquet/reader_impl_helpers.cpp [hunk 3]
```diff
@@ -346,6 +380,147 @@ size_type aggregate_reader_metadata::calc_num_row_groups() const
...
+          pg_info.num_valid     = num_values - pg_info.num_nulls.value();
+        }
+      }
+
+      // If none of the ifs above triggered, then we have neither histogram (likely the writer
+      // doesn't produce them, the r:0 d:1 case should have been handled above). The column index
+      // doesn't give us value counts, so we'll have to rely on the page headers. If the histogram
+      // info is missing or insufficient, then just return without modifying the row_group_info.
+      if (not pg_info.num_nulls.has_value() or not pg_info.num_valid.has_value()) {
+        rg_info.columns = std::nullopt;
```
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/reader_impl_helpers.cpp:506]: Yessss....

File: cpp/src/io/parquet/reader_impl_chunking.cu (hunk 1 above)
Comment (etseidl) [reply cpp/src/io/parquet/reader_impl_chunking.cu:376]: In fact this is similar to what `_metadata->get_column_metadata(rg.index, rg.source_index, col.schema_idx)` is doing below.

Comment (vuule) [reply cpp/src/io/parquet/reader_impl_chunking.cu:376]: Trying to find a(n easy) way to avoid the quadratic complexity here. Should be fine as-is.

**Comment by etseidl** [comment]
Going to close this for now and wait for #14360 to merge before trying again.

**Pull request closed without merging**